x-pack/plugins/security_solution/server/endpoint/routes/metadata/support/unenroll.ts

```diff
@@ -4,113 +4,33 @@
-  hostId: string
-): Promise<HostId | undefined> {
-  const queryParams = {
-    index: metadataMirrorIndexPattern,
-    body: {
-      size: 1,
-      _source: ['host.id'],
-      query: {
-        bool: {
-          filter: [
-            {
-              term: {
-                'Endpoint.status': EndpointStatus.unenrolled,
-              },
-            },
-            {
-              term: {
-                'host.id': hostId,
-              },
-            },
-          ],
-        },
-      },
-    },
-  };
-
-  const response = (await client.callAsCurrentUser('search', queryParams)) as SearchResponse<
-    HostId
-  >;
-  const newHits = response.hits?.hits || [];
-
-  if (newHits.length > 0) {
-    const hostIds = newHits.map((hitSource: HitSource) => hitSource._source);
-    return hostIds[0];
-  } else {
-    return undefined;
-  }
-}
-
-export async function findAllUnenrolledHostIds(
-  client: ILegacyScopedClusterClient
-): Promise<HostId[]> {
-  const queryParams = {
-    index: metadataMirrorIndexPattern,
-    scroll: KEEPALIVE,
-    body: {
-      size: SIZE,
-      _source: ['host.id'],
-      query: {
-        bool: {
-          filter: {
-            term: {
-              'Endpoint.status': EndpointStatus.unenrolled,
-            },
-          },
-        },
-      },
-    },
+import { SavedObjectsClientContract } from 'kibana/server';
+import { AgentService } from '../../../../../../ingest_manager/server';
+import { Agent } from '../../../../../../ingest_manager/common/types/models';
+
+export async function findAllUnenrolledAgentIds(
+  agentService: AgentService,
+  soClient: SavedObjectsClientContract,
+  pageSize: number = 1000
+): Promise<string[]> {
+  const searchOptions = (pageNum: number) => {
+    return {
+      page: pageNum,
+      perPage: pageSize,
+      showInactive: true,
+      kuery: 'fleet-agents.packages:endpoint AND fleet-agents.active:false',
+    };
   };
-  const response = (await client.callAsCurrentUser('search', queryParams)) as SearchResponse<
-    HostId
-  >;
-
-  return fetchAllUnenrolledHostIdsWithScroll(response, client.callAsCurrentUser);
-}
-
-export async function fetchAllUnenrolledHostIdsWithScroll(
-  response: SearchResponse<HostId>,
-  client: LegacyAPICaller,
-  hits: HostId[] = []
-): Promise<HostId[]> {
-  let newHits = response.hits?.hits || [];
-  let scrollId = response._scroll_id;
 
-  while (newHits.length > 0) {
-    const hostIds: HostId[] = newHits.map((hitSource: HitSource) => hitSource._source);
-    hits.push(...hostIds);
+  let page = 1;
 
-    const innerResponse = await client('scroll', {
-      body: {
-        scroll: KEEPALIVE,
-        scroll_id: scrollId,
-      },
-    });
+  const result: string[] = [];
+  let hasMore = true;
 
-    newHits = innerResponse.hits?.hits || [];
-    scrollId = innerResponse._scroll_id;
+  while (hasMore) {
+    const unenrolledAgents = await agentService.listAgents(soClient, searchOptions(page++));
```

Comment:
Will `listAgents` ever return `undefined`?

Reply:
as per @nchaulet it always return and empty list. And also tested. But I can be more defensive if needed.

Reply:
Naa, I just wanted to double check 👍